[RNMobile] Hide help button from UBE

SiobhyB opened the pull request:
Addresses https://github.com/wordpress-mobile/gutenberg-mobile/issues/4339

## Related PRs and Installable Builds

* `Gutenberg Mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/4352
* `WPiOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/17644
* `WPAndroid`: https://github.com/wordpress-mobile/WordPress-Android/pull/15674

## Description

As described in https://github.com/wordpress-mobile/gutenberg-mobile/issues/4339, a help button is currently displaying in the bottom right-hand corner of the unsupported block editor (UBE) in both the Android and iOS apps. The button can be seen while logged into WordPress.com. When tapped on, the button isn't working as expected on iOS and prompts users to navigate away from the editor on Android, creating the potential for edits to be lost.

With this PR, the button is hidden from view in order to avoid confusion and potential loss of edits. A high-level overview of the code changes can be found in the `Types of changes` section below.

## How has this been tested?

While logged into a WordPress.com site:

* Using a web browser, add a block that's not supported by Gutenberg Mobile (currently the Markdown block will work).
* Save the draft or publish the post.
* Load the post in the Gutenberg Mobile editor on WPiOS or WPAndroid.
* Choose the option to "Edit using web editor" on the unsupported block.
* Confirm that the help button no longer displays in the bottom right-hand corner of the UBE.

Installable builds have been created for [WPiOS here](https://github.com/wordpress-mobile/WordPress-iOS/pull/17644#issuecomment-988776948) and [WPAndroid here](https://github.com/wordpress-mobile/WordPress-Android/pull/15674#issuecomment-988796513) to help with testing.

## Screenshots <!-- if applicable -->

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/145207216-9f3a7f60-93c6-4aaf-bcd1-0ffcf3d306a6.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/145227468-d21c4c94-7def-4d6a-919c-7efd6d1e855e.png" width="100%"> |

## Types of changes

Bug fix (non-breaking change which fixes an issue), with the following high-level overview of the code changes involved:

### Gutenberg Mobile changes

* A new file, `external-style-overrides.css`, has [been created here](https://github.com/wordpress-mobile/gutenberg-mobile/blob/85070711a4cf92b98c4fc3578a5f2945b2ee6288/resources/unsupported-block-editor/external-style-overrides.css). This new file lives alongside other external/custom scripts [in Gutenberg Mobile's `unsupported-block-editor` folder](https://github.com/wordpress-mobile/gutenberg-mobile/tree/85070711a4cf92b98c4fc3578a5f2945b2ee6288/resources/unsupported-block-editor). Its purpose will be to house CSS that changes parts of the editor that are external to Gutenberg e.g. buttons added by third-party plugins.
* [CSS to hide a help button](https://github.com/wordpress-mobile/gutenberg-mobile/blob/85070711a4cf92b98c4fc3578a5f2945b2ee6288/resources/unsupported-block-editor/external-style-overrides.css#L1) that's added by Jetpack/WP.com (commonly used third-parties amongst app users) has been added to the new CSS file.
* I've dived into more detail around the iOS and Android changes below. For both platforms, I've attempted to apply the same patterns I found used for the existing JS files in Gutenberg Mobile's `unsupported-block-editor` folder, but tweaked for the newly added CSS file.

### Android changes

* A `getOnGutenbergReadyExternalStyles` function that [fetches the contents of the newly created CSS file](https://github.com/wordpress-mobile/WordPress-Android/blob/61ddfd081f1231ba46230197e433c69220ea047f/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/WPGutenbergWebViewActivity.java#L141-L144) has been added to Android's `WPGutenbergWebViewActivity` class.
* Any external styles defined in the new CSS file will now be [processed on the Gutenberg side](https://github.com/WordPress/gutenberg/blob/a2b298a47e8ecb93bf504ca70ddf75c886a24246/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergWebViewActivity.java#L337-L339) by first calling `getOnGutenbergReadyExternalStyles` and then following the same steps as the other internal CSS files.

### iOS changes

* We previously only referenced `.js` files in Gutenberg Mobile's `unsupported-block-editor` folder. As this PR adds a `.css` file to that folder, it was necessary to add `css` to [the global pattern here](https://github.com/wordpress-mobile/gutenberg-mobile/blob/bcb75bd442f453716914c59427896ef7f1a058f8/Gutenberg.podspec#L21) in the Gutenberg Mobile repository
* In the iOS repository's [GutenbergWebViewController.swift file](https://github.com/wordpress-mobile/WordPress-iOS/blob/a09fbb771a0d5e78322172bbe94139352a09f9fe/WordPress/Classes/ViewRelated/Gutenberg/GutenbergWeb/GutenbergWebViewController.swift#L109), a `onGutenbergReadyStyles` function calls `loadCustomStyles()`, which then [fetches the contents of the newly created CSS file](https://github.com/wordpress-mobile/WordPress-iOS/blob/a09fbb771a0d5e78322172bbe94139352a09f9fe/WordPress/Classes/ViewRelated/Gutenberg/GutenbergWeb/GutenbergWebViewController.swift#L109). This follows the same pattern as, [loadCustomScript()](https://github.com/wordpress-mobile/WordPress-iOS/blob/a09fbb771a0d5e78322172bbe94139352a09f9fe/WordPress/Classes/ViewRelated/Gutenberg/GutenbergWeb/GutenbergWebViewController.swift#L100), the function used to process `.js` files.
* In order to replicate the way other CSS files are processed in the Gutenberg repository, it was necessary to make [`getContent()`](https://github.com/WordPress/gutenberg/blob/95929b58e4f3a6bdc3f2df3cb612858d5051952b/packages/react-native-bridge/ios/SourceFile.swift#L23)  and [`toJsScript`](https://github.com/WordPress/gutenberg/blob/95929b58e4f3a6bdc3f2df3cb612858d5051952b/packages/react-native-bridge/ios/GutenbergWebFallback/FallbackJavascriptInjection.swift#L71) public for use in the new `loadCustomStyles()` function.
* Finally, `onGutenbergReadyStyles()` is called in all the same places as `injectEditorCssScript()` in [the GutenbergWebSingleBlockViewController.swift file](https://github.com/WordPress/gutenberg/blob/95929b58e4f3a6bdc3f2df3cb612858d5051952b/packages/react-native-bridge/ios/GutenbergWebFallback/GutenbergWebSingleBlockViewController.swift#L170). 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Hide help button from UBE

With this commit, the help button that currently displays in the bottom-right corner of the UBE is hidden via CSS.

SiobhyB committed:
Remove external CSS

As this CSS refers to an external third-party plugin, it's not appropriate to reference it in the Gutenberg repository. It'll therefore be removed from this PR and moved to a separate repository.

SiobhyB committed:
Inject CSS styles from external sources

With this commit, 'getOnGutenbergReadyExternalStyles' is called to fetch any external CSS styles that have been defined e.g. third-party plugins may need to customize certain parts of the UBE.

SiobhyB committed:
Correct 'externalStyles' function access modifier

'getOnGutenbergReadyExternalStyles' had been mistakenly set as 'private', even though it's called outside of thid specific class. By setting to 'protected', it can now be called in the 'WPGutenbergWebViewActivity' class, as expected.

SiobhyB committed:
Enable external styles to be added to Gutenberg on iOS-side

The 'onGutenbergReadyStyles' function will enable developers to add external CSS files to the UBE on the native iOS-side.

The pattern used to implement the 'onGutenbergReadyScripts' function (which enables external scripts) was followed.

SiobhyB committed:
Modify access permissions for necessary functions

getContent() and toJsScript() are needed to process external CSS files. Therefore, this commit modifies their access permissions so that they can be used outside of their respective classes.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/hide-help-button-from-ube

twstokes commented on packages/react-native-bridge/ios/GutenbergWebFallback/GutenbergWebSingleBlockViewController.swift:
```diff
@@ -152,6 +158,7 @@ extension GutenbergWebSingleBlockViewController: WKNavigationDelegate {
         evaluateJavascript(jsInjection.injectWPBarsCssScript)
         evaluateJavascript(jsInjection.injectLocalStorageScript)
         onPageLoadScripts().forEach(evaluateJavascript)
+        onGutenbergReadyStyles().forEach(evaluateJavascript)
```
👋 @SiobhyB - I notice that we're calling `onGutenbergReadyStyles` on both `didCommit` and `didFinish` for the `WKNavigationDelegate` - do you think perhaps we can get away with just using `didCommit`, or did you find `didFinish` was also necessary?

From my understanding, scripts are loaded in this sequence:

1. `didCommit` for the navigation delegate (this line)
2. `didFinish` for the navigation delegate
3. `onGutenbergReady()`

**This also brings up another question I wanted to ask:**

Should we rename `onGutenbergReadyStyles()` to something that matches the timing of when the script is loaded? I'm thinking this CSS injection works best when the page is loaded, but before it's ready. Some ideas I had were `onPageLoadStyles()` or `onGutenbergLoadStyles()`, but I'd love to hear your thoughts. Thanks!

SiobhyB replied:
👋 @twstokes, thank you for your help here! 🙇‍♀️ 

> I notice that we're calling `onGutenbergReadyStyles` on both `didCommit` and `didFinish` for the `WKNavigationDelegate` - do you think perhaps we can get away with just using `didCommit`, or did you find `didFinish` was also necessary?

I had a feeling both were necessary as I saw `injectEditorCssScript` (which loads the CSS from [this file](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/common/gutenberg-web-single-block/editor-style-overrides.css)) was also called from both, with [the following comment](https://github.com/WordPress/gutenberg/blob/2b18866215a19f87f40d59b275e43b6b7b2b9781/packages/react-native-bridge/ios/GutenbergWebFallback/GutenbergWebSingleBlockViewController.swift#L165-L166) alongside `didFinish`: 

```
// Sometimes the editor takes longer loading and its CSS can override what
// Injectic Editor specific CSS when everything is loaded to avoid overwritting parameters if gutenberg CSS load later.
```

My thinking is that the CSS from `onGutenbergReadyStyles` serves basically the same purposes as the CSS from `injectEditorCssScript` (to hide or customize parts of the editor), except that its housed under the Mobile Gutenberg repository. So, I tried to follow the same patterns as `injectEditorCssScript`.

Does that make sense to you? Or do you think there's a case for removing `onGutenbergReadyStyles` from `didFinish`?

> Should we rename onGutenbergReadyStyles() to something that matches the timing of when the script is loaded? I'm thinking this CSS injection works best when the page is loaded, but before it's ready. Some ideas I had were onPageLoadStyles() or onGutenbergLoadStyles(), but I'd love to hear your thoughts. Thanks!

I think renaming makes sense! I like `onGutenbergLoadStyles`, as I feel like it'd work for both the `didCommit` and `didFinish` scenarios (if we do continue to call `onGutenbergLoadStyles` from both). I've gone ahead to update that in https://github.com/WordPress/gutenberg/pull/37221/commits/e8371aaadad36353c7cce82ef6d93056611b2d5b. Looking forward to hearing what you think!

twstokes replied:
> Does that make sense to you?

Sure does, and it won't hurt a thing to keep it. Sounds good to me to keep the patterns consistent @SiobhyB!

twstokes commented on packages/react-native-bridge/ios/GutenbergWebFallback/FallbackJavascriptInjection.swift:
```diff
@@ -68,11 +68,13 @@ public extension SourceFile {
 }
 
-internal extension String {
+public extension String {
```
We _may_ be able to revert this back to `internal`, because I had a thought about how we're calling the `jsScript() -> WKUserScript` function:

I noticed in the [WPiOS PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/17644/files#diff-4aadfc0278d3b8272613aa3de61efa1636c9d9bf943f5346e3ebb98b547098c2R112) that we used a string literal to construct the `window.injectCss` call and I'm thinking about potential pitfalls in the future if someone were to rename that injected function.

An idea could be to take the [`SourceFile` extension](https://github.com/WordPress/gutenberg/pull/37221/files#diff-ec8720728c732299da7f328b24179bbbde3596fab310df48b6ffe64938b6356bL63) and move it to [`SourceFile.swift`](https://github.com/WordPress/gutenberg/blob/95929b58e4f3a6bdc3f2df3cb612858d5051952b/packages/react-native-bridge/ios/SourceFile.swift) where we could access its private variables and dynamically generate the necessary JavaScript depending on its type.

```
extension SourceFile {
    public func jsScript(with argument: String? = nil) throws -> WKUserScript {
        let content = try getContent()
        let formatted = String(format: content, argument ?? [])
        
        switch self.type {
        case .css:
            let injectCssScriptTemplate = "window.injectCss(`%@`)"
            return String(format: injectCssScriptTemplate, formatted).toJsScript()
        case .js, .json:
            return formatted.toJsScript()
        }
    }
}
```

This could keep the injection template string tightly coupled with the `inject-css` file, as well as simplify the `loadCustomStyles` call in the [WPiOS PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/17644/files#diff-4aadfc0278d3b8272613aa3de61efa1636c9d9bf943f5346e3ebb98b547098c2R111) to something like: `return try SourceFile(name: name, type: .css).jsScript(with: argument)`

I'm also OK with how it is now, so I wanted to get your thoughts. 🙇

SiobhyB replied:
I really appreciate this insight into better programming practices, thank you 🙇‍♀️ What you've outlined all seems 
favourable to me and I've gone ahead to implement it on the Gutenberg side in https://github.com/WordPress/gutenberg/pull/37221/commits/0cda8b3e569faba16bc10aa39c60d44658e92268. (I'll implement the necessary changes on the iOS side when I make my way to that review!)

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/hide-help-button-from-ube

SiobhyB committed:
Rename function to 'onGutenbergLoadStyles' for clarity

This commit renames 'onGutenbergLoadStyles' to more accurately reflect the timing of the script.

SiobhyB committed:
Refactor SourceFile extension

With this commit, the SourceFile extension is moved to 'SourceFile.swift' and refactored so that JavaScript is dynamically generated based on file type (e.g. CSS, JS, etc). This refactoring also means that we're now able to revert the change that was made to the 'toJsScript' function's access modifier.

This follows the suggestion here, which also further details the benefits of this approach: https://github.com/WordPress/gutenberg/pull/37221#discussion_r771093813

twstokes commented on packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergWebViewActivity.java:
```diff
@@ -333,10 +333,18 @@ private void injectCssScript() {
                 injectWPBarsCssScript = removeWhiteSpace(removeNewLines(injectWPBarsCssScript));
                 evaluateJavaScript(String.format(INJECT_CSS_SCRIPT_TEMPLATE, injectWPBarsCssScript));
+
+                String injectExternalCssScript = getOnGutenbergReadyExternalStyles();
```
This looks good and works well in my testing! I did have a thought on how iOS is injecting vs Android:

On iOS we're injecting the CSS when the [webview's page is first loaded](https://github.com/WordPress/gutenberg/pull/37221/files#diff-07dba60ae8d26048cd7620ab6004ed8298005beb404b40d804a580dc064814dbR161). If we'd waited for the `onGutenbergReady()` call the question mark icon would briefly be shown and then disappear because the CSS wouldn't be evaluated until everything was "ready".

On Android we're waiting for `onGutenbergReady()`, though I don't see the same effect where the icon is shown and disappears, so that's good. I don't know for sure if that will always be the case, though.

Do you think we should make the timing of both match to be consistent? For Android this would mean making the call at either [onPageCommitVisible](https://developer.android.com/reference/android/webkit/WebViewClient#onPageCommitVisible(android.webkit.WebView,%20java.lang.String)) or [onPageStarted](https://developer.android.com/reference/android/webkit/WebViewClient#onPageStarted(android.webkit.WebView,%20java.lang.String,%20android.graphics.Bitmap)), but we may want to summon an Android dev for the best advice. 😄

SiobhyB replied:
Thank you for testing and the review, @twstokes! 

> Do you think we should make the timing of both match to be consistent? For Android this would mean making the call at either onPageCommitVisible or onPageStarted, but we may want to summon an Android dev for the best advice. 😄

I followed the way [the existing `editor-style-overrides.css` file](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/common/gutenberg-web-single-block/editor-style-overrides.css) is loaded for both platforms as a pattern (i.e. under [this code block](https://github.com/WordPress/gutenberg/blob/0cda8b3e569faba16bc10aa39c60d44658e92268/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergWebViewActivity.java#L325-L342) on Android) under the assumption it's best for the new `external-style-overrides.css` file to load at the same time as that one. If we loaded one or the other at separate times, I worry that there could be cases where parts of the editor are hidden at differing times. WDYT?

Maybe there's a case for creating a separate issue to address the inconsistency for how UBE styles are loaded across platforms?

twstokes replied:
That makes sense @SiobhyB - so, basically we're following patterns for their respective platforms. In that case that's all I had!

twstokes commented:
I followed the testing steps on both iOS and Android they passed! 🎉

twstokes reviewed (approved):
LGTM @SiobhyB. 👍

SiobhyB committed:
Merge branch 'trunk' into rnmobile/update/hide-help-button-from-ube

SiobhyB committed:
Update CHANGELOG
